src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,333 @@
+
+        f = float(1 / p)  # the instrument accepts frequency as input so period is inverted
+        self._sendCommand("sour%d:freq %f" %(c, f))
+
+    def setVoltageMin(self, c, v):
+        """
+        c (int): the channel to set the low voltage. 1 or 2
+        v (int): the voltage value
+        """
+        self._sendCommand("sour%d:volt:low %d" % (c, v))
+
+    def setVoltageMax(self, c, v):
+        """
+        c (int): the channel to set the period of. 1 or 2
+        p (str): the period value
+        """
+        self._sendCommand("sour%d:volt:high %d" % (c, v))
+
+    def setVoltageLimitMin(self, c, v):
+        """
+        c (int): the channel to set the period of. 1 or 2
+        p (str): the period value
+        """
+        self._sendCommand("sour%d:volt:lim:low %d" % (c, v))
+
+    def setVoltageLimitMax(self, c, v):
+        """
+        c (int): the channel to set the period of. 1 or 2
+        p (str): the period value
+        """
+        self._sendCommand("sour%d:volt:lim:high %d" % (c, v))
+
+    def setDutyCycle(self, c, dc):
+        """
+        c (int): the channel to set the period of. 1 or 2
+        p (str): the period value
+        """
+        self._sendCommand("sour%d:func:squ:dcyc %s" % (c, dc))
+
+    def setTracking(self, c, t):
+        """
+        c (int): the channel to set the period of. 1 or 2
+        p (str): the period value
+        """
+        self._sendCommand("sour%d:trac %f" % (c, t))
+
+
+class IPBusAccesser(object):
+    """
+    Manage TCP/IP connections over ethernet
+    """
+
+    def __init__(self, ip_addr=None, tcp_timeout=10.0, tcp_port=23):
+        """ Initialize this `TEWWAwg` instance.
+
+        If the given `ip_addr` is not None, then communication is open.
+
+        :param ip_addr: the instrument's IP-Address (digits and dots format)
+        :param tcp_timeout: TCP-Socket time-out (in seconds)
+        """
+        self._tcp_sock = None
+        self._tcp_port = tcp_port
+        self._ip_addr = ip_addr
+        self._tcp_timeout = float(tcp_timeout)
+
+        if ip_addr is not None:
+            self.open()
+
+    def open(self, ip_addr=None, tcp_timeout=None):
+        """Open/Reopen Connection
+
+        If a given argument is None, the previous value of the same argument is used.
+
+        :param ip_addr: the instrument's IP-Address (digits and dots format)
+        :param tcp_timeout: TCP-Socket time-out (in seconds)
+        """
+        self.close()
+
+        if ip_addr is not None:
+            self._ip_addr = ip_addr
+        if tcp_timeout is not None:
+            self._tcp_timeout = float(tcp_timeout)
+
+        # Open TCP-IP Socket:
+        self._tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
+        self._tcp_sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        self._tcp_sock.settimeout(self._tcp_timeout)
+        self._tcp_sock.connect((self._ip_addr, self._tcp_port))
```

Comment:
You have to handle the case when the IP address cannot be found. Either here, or in the caller. In such case, it should raise a `HwError` to tell the Odemis back-end that a non-fatal error happened (and pass a user-friendly message asking to check the device is connected and turned on).